{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "54a85a1a-e636-4961-bd37-ae3792a93ee8",
   "metadata": {},
   "source": [
    "# NMD annotations\n",
    "This script labels genomic positions with an NMD annotation. The annotations include:\n",
    "1) Start-proximal (<150nt from translation start site)\n",
    "2) Long exons (>400nt upstream of the splice donor site)\n",
    "3) Last exon\n",
    "4) 50nt rule (within the most 3' 50nt of the penultimate exon)\n",
    "\n",
    "There are several approaches I could take.\n",
    "First, I could annotate every position in the exome with these criteria.\n",
    "Alternatively, I could construct a bed file of different regions.\n",
    "\n",
    "Generally, I would favour the per-site annotations for maximum flexibility.\n",
    "So I will need to annotate:\n",
    "1) The distance from the translation start site\n",
    "2) The distance upstream from the splice donor site\n",
    "3) Whether it is the last exon\n",
    "4) Whether it is the last 50nt of the penultimate exon"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccacf2cf-846a-4cfa-b506-55ed0d0a36a7",
   "metadata": {},
   "source": [
    "## Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a50eac1-d44c-4d10-8685-15aa62e9b0f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda install -c bioconda gtfparse -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ea1ee56-91b6-4f96-8846-861e324d7bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "dx download -o ../data/ data/gencode.v39.annotation.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "eb1a4a27-f37d-4b3a-b5b0-14994d267f9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the relevant modules\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gtfparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b0238b0f-b902-4095-837b-c6ce766ed80e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gencode_gtf(path):\n",
    "    \"\"\"Read a GENCODE .gtf into memory with gtfparse\"\"\"\n",
    "    gtf = gtfparse.read_gtf(path)\n",
    "    return gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ee77310b-b02e-4ec5-b063-ec552a406b09",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_canonical_cds(gtf):\n",
    "    \"\"\"Identify all CDS features in each Ensembl_canonical from GENCODE\"\"\"\n",
    "\n",
    "    # Subset to Ensembl_canonical CDS features in protein coding genes\n",
    "    canonical_cds = gtf[\n",
    "        (gtf.feature == \"CDS\")\n",
    "        & (gtf.tag.str.contains(\"Ensembl_canonical\"))\n",
    "        & (gtf.gene_type == \"protein_coding\")\n",
    "    ].copy()\n",
    "\n",
    "    # Count the number of CDS exons in each MANE transcript\n",
    "    canonical_cds[\"exon_number\"] = canonical_cds[\"exon_number\"].astype(int)\n",
    "    canonical_cds[\"cds_number\"] = (\n",
    "        canonical_cds.groupby(\"transcript_id\")[\"exon_number\"].rank().astype(int)\n",
    "    )\n",
    "    return canonical_cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "24d1b858-4eb9-410c-b25d-dc694ef3f53e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ablakes/.conda/envs/termini/lib/python3.10/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/ablakes/.conda/envs/termini/lib/python3.10/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65564</td>\n",
       "      <td>65573</td>\n",
       "      <td>ENSG00000186092.7,ENST00000641515.2,ENSE000038...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69036</td>\n",
       "      <td>70005</td>\n",
       "      <td>ENSG00000186092.7,ENST00000641515.2,ENSE000038...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450742</td>\n",
       "      <td>451678</td>\n",
       "      <td>ENSG00000284733.2,ENST00000426406.4,ENSE000023...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685718</td>\n",
       "      <td>686654</td>\n",
       "      <td>ENSG00000284662.2,ENST00000332831.5,ENSE000023...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924431</td>\n",
       "      <td>924948</td>\n",
       "      <td>ENSG00000187634.13,ENST00000616016.5,ENSE00001...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3240772</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209531</td>\n",
       "      <td>57209733</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3240774</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209821</td>\n",
       "      <td>57209980</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3240776</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210639</td>\n",
       "      <td>57210792</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3240778</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211551</td>\n",
       "      <td>57211620</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3240780</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211760</td>\n",
       "      <td>57211891</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196885 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname     start       end  \\\n",
       "61         chr1     65564     65573   \n",
       "64         chr1     69036     70005   \n",
       "322        chr1    450742    451678   \n",
       "449        chr1    685718    686654   \n",
       "943        chr1    924431    924948   \n",
       "...         ...       ...       ...   \n",
       "3240772    chrY  57209531  57209733   \n",
       "3240774    chrY  57209821  57209980   \n",
       "3240776    chrY  57210639  57210792   \n",
       "3240778    chrY  57211551  57211620   \n",
       "3240780    chrY  57211760  57211891   \n",
       "\n",
       "                                                        id score strand  \n",
       "61       ENSG00000186092.7,ENST00000641515.2,ENSE000038...     .      +  \n",
       "64       ENSG00000186092.7,ENST00000641515.2,ENSE000038...     .      +  \n",
       "322      ENSG00000284733.2,ENST00000426406.4,ENSE000023...     .      -  \n",
       "449      ENSG00000284662.2,ENST00000332831.5,ENSE000023...     .      -  \n",
       "943      ENSG00000187634.13,ENST00000616016.5,ENSE00001...     .      +  \n",
       "...                                                    ...   ...    ...  \n",
       "3240772  ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...     .      +  \n",
       "3240774  ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...     .      +  \n",
       "3240776  ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...     .      +  \n",
       "3240778  ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...     .      +  \n",
       "3240780  ENSG00000182484.15_PAR_Y,ENST00000359512.8_PAR...     .      +  \n",
       "\n",
       "[196885 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # Read GTF data\n",
    "    gencode_path = \"../data/gencode.v39.annotation.gtf\"\n",
    "    gtf = get_gencode_gtf(gencode_path)\n",
    "\n",
    "    # Define regions of interest\n",
    "    cds = get_canonical_cds(gtf)\n",
    "    cds = cds[[\"seqname\",\"start\",\"end\",\"strand\",\"transcript_id\",\"exon_id\",\"exon_number\",\"cds_number\"]]\n",
    "    cds = cds.set_index([\"seqname\",\"transcript_id\",\"exon_id\"])\n",
    "    cds[\"pos\"] = cds.apply(lambda x: list(range(x[\"start\"], x[\"end\"] + 1)), axis=1)\n",
    "    cds = cds.explode(\"pos\")\n",
    "    cds[\"cds_len\"] = cds.groupby([\"transcript_id\"])[\"pos\"].count()\n",
    "    \n",
    "    # Convert .gtf data to .bed format\n",
    "    bed_ids = [\"gene_id\", \"transcript_id\", \"exon_id\", \"cds_number\"]\n",
    "    cds_bed = gtf_to_bed(cds, bed_ids)\n",
    "\n",
    "    # Write to output\n",
    "    gencode_version = \"v39\"\n",
    "    feature = \"cds\"\n",
    "    out_path_chr = f\"../outputs/gencode_{gencode_version}_canonical_{feature}_chr.bed\"\n",
    "\n",
    "    write_bed(cds_bed, out_path_chr, \"chr\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
